src/core/python2js_buffer.js

```diff
@@ -0,0 +1,214 @@
+                      `got '${formatStr}'.` + errorMessage);
+    }
+    let formatChar = formatStr.slice(-1);
+    let alignChar = formatStr.slice(0, -1);
+    let bigEndian;
+    switch (alignChar) {
+    case "!":
+    case ">":
+      bigEndian = true;
+      break;
+    case "<":
+    case "@":
+    case "=":
+    case "":
+      bigEndian = false;
+      break;
+    default:
+      throw new Error(`Unrecognized alignment character ${alignChar}.` +
+                      errorMessage);
+    }
+    let arrayType;
+    switch (formatChar) {
+    case 'b':
+      arrayType = Int8Array;
+      break;
+    case 's':
+    case 'p':
+    case 'c':
+    case 'B':
+    case '?':
+      arrayType = Uint8Array;
+      break;
+    case 'h':
+      arrayType = Int16Array;
+      break;
+    case 'H':
+      arrayType = Uint16Array;
+      break;
+    case 'i':
+    case 'l':
+    case 'n':
+      arrayType = Int32Array;
+      break;
+    case 'I':
+    case 'L':
+    case 'N':
+    case 'P':
+      arrayType = Uint32Array;
+      break;
+    case 'q':
+      // clang-format off
+            if (globalThis.BigInt64Array === undefined) {
+        // clang-format on
+        throw new Error("BigInt64Array is not supported on this browser." +
+                        errorMessage);
+      }
+      arrayType = BigInt64Array;
+      break;
+    case 'Q':
+      // clang-format off
+            if (globalThis.BigUint64Array === undefined) {
+        // clang-format on
+        throw new Error("BigUint64Array is not supported on this browser." +
+                        errorMessage);
+      }
+      arrayType = BigUint64Array;
+      break;
+    case 'f':
+      arrayType = Float32Array;
+      break;
+    case 'd':
+      arrayType = Float64Array;
+      break;
+    case "e":
+      throw new Error("Javascript has no Float16 support.");
+    default:
+      throw new Error(`Unrecognized format character '${formatChar}'.` +
+                      errorMessage);
+    }
+    return [ arrayType, bigEndian ];
+  };
+
+  Module.python2js_buffer_1d_contiguous = function(ptr, stride, n, converter) {
+    "use strict";
+    let byteLength = stride * n;
+    let backing = HEAP8.slice(ptr, ptr + byteLength).buffer;
+    return converter(backing);
+  };
+
+  Module.python2js_buffer_1d_noncontiguous = function(ptr, stride, suboffset, n,
+                                                      itemsize, converter) {
+    "use strict";
+    let byteLength = itemsize * n;
+    let buffer = new Uint8Array(byteLength);
+    for (i = 0; i < n; ++i) {
+      let curptr = ptr + i * stride;
+      if (suboffset >= 0) {
+        curptr = HEAP32[curptr / 4] + suboffset;
+      }
+      buffer.set(HEAP8.subarray(curptr, curptr + itemsize), i * itemsize);
+    }
+    return converter(buffer.buffer);
+  };
+
+  Module._python2js_buffer_recursive = function(ptr, curdim, bufferData) {
+    "use strict";
+    let n = HEAP32[bufferData.shape / 4 + curdim];
```

Comment:
Please add an inline comment why we need the "/4" (for adaptation from HEAP8 to HEAP32 I imagine?)